swingset/src/main/java/com/nqadmin/swingset/SSDataNavigator.java

```diff
@@ -902,7 +902,12 @@ private boolean commitChanges() throws SQLException {
 			}
 			//rowSet.updateRow();
 			//dBNav.performPostUpdateOps();
-			doUpdateRow();
+
+			// 2021-11-18: prasanthreddy-git noticed that we didn't check for an empty
+			// rowset before trying to save changes
+			if (rowSet.getRow() > 0) {
+				doUpdateRow();
+			}
```

Comment:
Looks like we are calling performPostUpdateOps in doUpdateRow which should not apply if the update is not needed. So either place should be fine. Would be better if we bring the call to rowset update & postUpdateOps itself into commitChanges as doUpdateRowWithoutPostUpdateOps & doUpdateRow don't have any logic in them.

Reply:
(the logic I was referring to was calling postUpdateOps)

There was a proposed change for smart NavBar, see #40, that didn't make it into 4.0.

Some of the changes from that proposal were brought in to 4.0; particularly those that localized setting the enable/disable state of the NavBar buttons rather than having that logic scattered around the actions since that tended to be buggy. While doing that the method `commitChanges()` was introduced which subsumed `doUpdateRow()`. So yeah, now I agree that the _fix belongs in, or around, `commitChanges()`_ and **`doUpdateRow()` can be removed**.

I'd be inclined to change the name of `commitChanges()` to `commitChangesIfAllowed()`; with the name change, the conditional return which always appears when calling `commitChanges()` is easier to understand. And I'd probably change `doUpdateRowWithoutPostUpdateOps()` to something like `commitChangesIfAllowedWithoutPostUpdateOps()`, maybe a little less verbose but to something so the update methods sort together. With these name changes, the recommended fix would go in `commitChangesIfAllowedWithoutPostUpdateOps()`.

Reply:
Let me take a stab at this and commit an updated SSDataNavigator

Reply:
@errael  @prasanthreddy-git  Have a look at the SSDataNavigator changes (pushed to this merge request) and see if they are to your liking. Tested a bit and seems to be working as expected. Hopefully code is a bit cleaner now.